### PR TITLE
make `.stash(..)` work in macros without ICEing

### LINKED ISSUE #69396: internal compiler error: missing type for `const` item
I'm seeing an internal compiler error on the following input (found by [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc)):

```rust
macro_rules! suite {
    ( $( $fn:ident; )* ) => {
        $(
            const A = "A".$fn();
        )*
    }
}

suite! {
    len;
    is_empty;
}

fn main() {}
```

```
$ rustc src/main.rs
error: internal compiler error: missing type for `const` item
  --> src/main.rs:4:19
   |
4  |               const A = "A".$fn();
   |                     ^ help: provide a type for the item: `A: <type>`
...
9  | / suite! {
10 | |     len;
11 | |     is_empty;
12 | | }
   | |_- in this macro invocation
   |
   = note: src/librustc_errors/lib.rs:456: already existing stashed diagnostic with (span = Span { lo: BytePos(80), hi: BytePos(81), ctxt: #3 }, key = ItemNoType)
   = note: this error: internal compiler error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:461:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.0-nightly (7760cd0fb 2020-02-19) running on x86_64-unknown-linux-gnu

error: missing type for `const` item
  --> src/main.rs:4:19
   |
4  |               const A = "A".$fn();
   |                     ^ help: provide a type for the item: `A: <type>`
...
9  | / suite! {
10 | |     len;
11 | |     is_empty;
12 | | }
   | |_- in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

```

### PR DESCRIPTION
**Pull request by Centril**
Fixes https://github.com/rust-lang/rust/issues/69396, which was injected by https://github.com/rust-lang/rust/pull/64698, by unique-ifying the `Span` part of the key used when stashing the diagnostic such that the key's slot isn't overwritten in `handler.stash_diagnostic(..)`.

This is similar to the usage of `fresh_expansion` in `mark_span_with_reason` (but that's not used for unique-ifying). Stashing is somewhat similar in the sense that it is sort of about compiler generated code as well.

An alternative approach would be to store `Vec<Diagnostic>` instead of `Diagnostic`, but the disadvantage of such a solution is that it might handle non-determinism poorly when stealing diagnostics (e.g. if stealing order would be different than stashing order). The approach in this PR feels more robust in comparison.

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
:cold_sweat:

**Comment by Centril** [comment]
That bad, eh? 😂

File: src/librustc_span/hygiene.rs [hunk 1]
```diff
@@ -729,6 +729,8 @@ pub enum ExpnKind {
     AstPass(AstPass),
     /// Desugaring done by the compiler during HIR lowering.
     Desugaring(DesugaringKind),
+    /// AST fragements produced by parser recovery.
```
Comment (spastorino) [inline src/librustc_span/hygiene.rs]:
```suggestion
    /// AST fragments produced by parser recovery.
```

**Comment by petrochenkov** [comment]
I'd really like to avoid introducing new uses of `fresh_expansion`, especially for something that is not expansion/desugaring.

**Comment by petrochenkov** [comment]
The diagnostic stashing problem can be generalized as "Give me a unique ID at point A that can be retrieved later at point B".

The both attempts at stashing API tried to use some existing data as dual purpose - their primary meaning and the unique ID.

The first attempt was to use `Span` as the unique ID, obviously a bad idea, with proc macros you can have a whole wall of code having the same macro invocation span.

The second attempt is to create unique `Span`s using macro expansion machinery.
This gives better uniqueness, but cannot be used with e.g. identifier spans which need their original spans for name resolution. It's also a pretty big hack in the sense of using tools existing for entirely different purposes.

**Comment by petrochenkov** [comment]
"Proper" solution would be using a separate ID, `DiagStashId` or something.
These IDs would be added to nodes using diagnostic stashing, like `ast::Ty` or what else is uses it currently.
Is it an overkill? Maybe yes.

Another alternative is to hide these stash IDs in `NodeId`s instead of spans.
- If the node ID is already assigned, it can be used for stashing.
- If the node ID is not yet assigned it can be
    - Assigned immediately (not sure whether the necessary infrastructure is ready, plus some extra IDs eliminated during expansion can be created, plus some validators may complain about non-contiguous ID ranges or something.)
   - Temporary ID can be assigned replaced by the proper ID (`fn visit_id` in `expand.rs`).
The proper node ID assignment can then update the "proper ID -> temporary ID" mapping, or update IDs in the diagnostic stash directly.

**Comment by petrochenkov** [comment]
I'd personally just nuke the whole diagnostic stashing and report everything in the traditional way :)
So far it didn't get any new uses since it was introduced.

**Comment by petrochenkov** [comment]
One more alternative: remove the assertion in `fn stash_diagnostic`.

If we hit this case, then we will report at least one error (assuming all stashed diagnostics are errors), so it's acceptable to omit some further errors.

This is perhaps the easiest way to fix https://github.com/rust-lang/rust/issues/69396 without introducing more evil into the codebase. (Also the most backportable.)

**Comment by Centril** [comment]
Some thoughts:

> I'd really like to avoid introducing new uses of fresh_expansion, especially for something that is not expansion/desugaring.

It might be a good idea to improve the docs on `fresh_expansion` to discourage new uses or something. (In general, the docs on `Span` & friends are pretty sparse.)

> The second attempt is to create unique `Span`s using macro expansion machinery.
> This gives better uniqueness, but cannot be used with e.g. identifier spans which need their original spans for name resolution. It's also a pretty big hack in the sense of using tools existing for entirely different purposes.

Could you elaborate a bit regarding name resolution? I don't think I understand the problem right now.

>  Is it an overkill? Maybe yes.

Yeah it does sound overkill -- the reason I used `Span` is because it was already used by almost everything, so I wouldn't need to make invasive AST structure changes.

> Another alternative is to hide these stash IDs in `NodeId`s instead of spans.

This does sound like a promising approach since `NodeId`s are also a ubiquitous structure in the AST data types. I don't think the HIR validator should be a problem since expansion should already be done before we reach lowering. You might not even need to replace temporary IDs at all with proper ones if we simply start the "proper" ones at +1 the highest stash-`NodeId`?

> I'd personally just nuke the whole diagnostic stashing and report everything in the traditional way :)
> So far it didn't get any new uses since it was introduced.

For what it's worth, I do have other cases in mind where I would like to use the stashing API but haven't gotten around to it yet. (I've mostly discussed these with @estebank in private.)

I would be OK with nuking the API if the consensus is that it is too architecturally problematic, though the diagnostics regressions would be unfortunate and increase pressure for something like https://github.com/rust-lang/rfcs/pull/2545. cc @rust-lang/wg-diagnostics 

> One more alternative: remove the assertion in `fn stash_diagnostic`.

Alright; I'll create another PR with this fix while we continue to discuss the long term strategy here.

**Comment by Centril** [comment]
Closing this PR in favor of https://github.com/rust-lang/rust/issues/70069 as this isn't a high priority item to work on for now.

**Pull request closed without merging**